# Lecture 2 — Class Exercise
# Bar Charts: World Happiness Report 2023

---

> **Your task:** Create **2 polished bar charts** using the World Happiness Report dataset.  
> **Push to:** `week02/lecture02_exercise.ipynb` in **your own GitHub repo** before the end of class.

---

### Rules (these will be checked in the model answer review next week)
1. Every bar chart **must have a zero baseline** — no exceptions (SWD p.51)
2. Every chart **must have an insight title**, not a topic title (SWD p.29)
3. Aim for **professional quality** — clean background, readable font, no clutter
4. Horizontal bars for long category names (SWD p.57)

---

## Setup — Run this cell first

In [1]:
import pandas as pd
import numpy as np
import plotly.graph_objects as go
import plotly.express as px
from plotly.subplots import make_subplots

template = "plotly_white"

# ─────────────────────────────────────────────────────────────────────────────
# World Happiness Report 2023 — embedded representative dataset
# Source: https://worldhappiness.report/ed/2023/
# ─────────────────────────────────────────────────────────────────────────────
data = [
    # Country, Region, Happiness_Score, GDP, Social_Support, Life_Expectancy, Freedom, Generosity, Corruption
    ("Finland","Western Europe",7.804,1.892,1.258,0.775,0.736,0.109,0.535),
    ("Denmark","Western Europe",7.586,1.953,1.243,0.783,0.719,0.233,0.564),
    ("Iceland","Western Europe",7.525,1.936,1.320,0.803,0.718,0.260,0.362),
    ("Israel","Western Europe",7.473,1.803,1.144,0.769,0.649,0.111,0.419),
    ("Netherlands","Western Europe",7.403,1.941,1.215,0.773,0.706,0.283,0.436),
    ("Sweden","Western Europe",7.395,1.908,1.225,0.785,0.752,0.253,0.523),
    ("Norway","Western Europe",7.315,2.027,1.243,0.790,0.746,0.262,0.496),
    ("Switzerland","Western Europe",7.240,2.025,1.268,0.807,0.727,0.226,0.484),
    ("Luxembourg","Western Europe",7.228,2.098,1.234,0.780,0.705,0.208,0.397),
    ("New Zealand","North America and ANZ",7.123,1.827,1.265,0.790,0.757,0.296,0.466),
    ("Austria","Western Europe",7.097,1.944,1.249,0.784,0.684,0.199,0.406),
    ("Australia","North America and ANZ",7.095,1.921,1.263,0.793,0.728,0.319,0.421),
    ("Canada","North America and ANZ",6.961,1.875,1.233,0.778,0.738,0.285,0.427),
    ("Ireland","Western Europe",6.911,2.093,1.264,0.786,0.714,0.218,0.408),
    ("United States","North America and ANZ",6.894,2.016,1.223,0.732,0.681,0.266,0.382),
    ("Germany","Western Europe",6.892,1.929,1.221,0.779,0.671,0.212,0.446),
    ("Belgium","Western Europe",6.859,1.922,1.227,0.786,0.654,0.170,0.380),
    ("Czech Republic","Central and Eastern Europe",6.845,1.808,1.250,0.773,0.655,0.067,0.420),
    ("United Kingdom","Western Europe",6.796,1.882,1.215,0.769,0.646,0.274,0.372),
    ("Lithuania","Central and Eastern Europe",6.763,1.819,1.244,0.729,0.688,0.049,0.338),
    ("France","Western Europe",6.661,1.879,1.201,0.784,0.618,0.130,0.332),
    ("Romania","Central and Eastern Europe",6.634,1.766,1.151,0.734,0.600,0.040,0.285),
    ("Slovenia","Central and Eastern Europe",6.530,1.855,1.215,0.787,0.683,0.078,0.352),
    ("Estonia","Central and Eastern Europe",6.527,1.840,1.221,0.765,0.679,0.040,0.393),
    ("Poland","Central and Eastern Europe",6.144,1.793,1.175,0.755,0.591,0.068,0.284),
    ("Uruguay","Latin America and Caribbean",6.494,1.749,1.234,0.756,0.687,0.131,0.396),
    ("Costa Rica","Latin America and Caribbean",6.410,1.644,1.183,0.767,0.722,0.161,0.373),
    ("Chile","Latin America and Caribbean",6.313,1.800,1.215,0.753,0.641,0.098,0.327),
    ("Brazil","Latin America and Caribbean",6.125,1.706,1.130,0.700,0.566,0.143,0.237),
    ("Argentina","Latin America and Caribbean",6.049,1.713,1.205,0.726,0.567,0.085,0.222),
    ("Mexico","Latin America and Caribbean",6.330,1.726,1.077,0.720,0.657,0.063,0.225),
    ("Colombia","Latin America and Caribbean",5.631,1.660,1.087,0.704,0.624,0.119,0.263),
    ("Peru","Latin America and Caribbean",5.395,1.658,1.099,0.688,0.601,0.101,0.181),
    ("Bolivia","Latin America and Caribbean",5.747,1.542,1.121,0.621,0.642,0.069,0.133),
    ("Singapore","Southeast Asia",6.587,2.127,1.171,0.820,0.655,0.380,0.455),
    ("Thailand","Southeast Asia",5.843,1.642,1.059,0.700,0.618,0.317,0.182),
    ("Malaysia","Southeast Asia",5.962,1.790,1.088,0.700,0.577,0.218,0.219),
    ("Philippines","Southeast Asia",5.994,1.474,1.071,0.609,0.679,0.237,0.184),
    ("Vietnam","Southeast Asia",5.773,1.551,1.167,0.685,0.661,0.188,0.143),
    ("Indonesia","Southeast Asia",5.240,1.583,1.177,0.643,0.619,0.261,0.221),
    ("Cambodia","Southeast Asia",4.968,1.378,0.949,0.614,0.614,0.260,0.165),
    ("Taiwan","East Asia",6.535,2.033,1.095,0.810,0.668,0.320,0.397),
    ("Japan","East Asia",6.129,1.952,1.249,0.835,0.611,0.176,0.434),
    ("South Korea","East Asia",5.951,1.948,1.157,0.776,0.540,0.213,0.323),
    ("China","East Asia",5.818,1.823,1.103,0.742,0.589,0.052,0.143),
    ("Mongolia","East Asia",5.285,1.592,1.181,0.620,0.617,0.144,0.141),
    ("Hong Kong","East Asia",5.437,2.166,1.155,0.834,0.499,0.239,0.368),
    ("Uzbekistan","Commonwealth of Independent States",6.179,1.421,1.283,0.614,0.782,0.406,0.148),
    ("Kazakhstan","Commonwealth of Independent States",5.765,1.836,1.243,0.657,0.643,0.070,0.142),
    ("Russia","Commonwealth of Independent States",5.661,1.855,1.220,0.635,0.348,0.068,0.072),
    ("Ukraine","Commonwealth of Independent States",5.071,1.686,1.154,0.683,0.601,0.119,0.211),
    ("Georgia","Commonwealth of Independent States",4.925,1.612,0.933,0.720,0.605,0.150,0.225),
    ("Armenia","Commonwealth of Independent States",5.399,1.614,1.110,0.673,0.607,0.100,0.192),
    ("Israel","Middle East and North Africa",7.473,1.803,1.144,0.769,0.649,0.111,0.419),
    ("Saudi Arabia","Middle East and North Africa",6.363,1.931,1.159,0.684,0.567,0.080,0.150),
    ("UAE","Middle East and North Africa",6.571,2.040,1.148,0.714,0.680,0.266,0.251),
    ("Kuwait","Middle East and North Africa",6.173,1.996,1.186,0.671,0.562,0.155,0.206),
    ("Morocco","Middle East and North Africa",5.298,1.499,0.846,0.634,0.538,0.107,0.171),
    ("Tunisia","Middle East and North Africa",5.174,1.526,0.953,0.668,0.406,0.068,0.128),
    ("Egypt","Middle East and North Africa",4.219,1.558,1.019,0.603,0.444,0.104,0.133),
    ("Jordan","Middle East and North Africa",4.633,1.512,0.990,0.672,0.511,0.134,0.171),
    ("Iraq","Middle East and North Africa",4.644,1.636,1.000,0.610,0.446,0.106,0.094),
    ("India","South Asia",4.036,1.434,0.869,0.541,0.553,0.199,0.166),
    ("Pakistan","South Asia",4.552,1.327,0.949,0.566,0.464,0.218,0.108),
    ("Bangladesh","South Asia",4.673,1.300,0.934,0.616,0.603,0.222,0.117),
    ("Nepal","South Asia",5.312,1.084,1.077,0.568,0.626,0.310,0.189),
    ("Sri Lanka","South Asia",4.388,1.575,0.999,0.713,0.579,0.213,0.091),
    ("Afghanistan","South Asia",1.859,0.350,0.476,0.361,0.000,0.158,0.025),
    ("Nigeria","Sub-Saharan Africa",5.119,1.367,0.894,0.349,0.559,0.212,0.122),
    ("Ghana","Sub-Saharan Africa",4.819,1.427,0.986,0.555,0.612,0.148,0.154),
    ("Kenya","Sub-Saharan Africa",4.580,1.363,1.018,0.476,0.574,0.237,0.136),
    ("Ethiopia","Sub-Saharan Africa",4.087,0.889,1.097,0.399,0.556,0.245,0.207),
    ("Tanzania","Sub-Saharan Africa",3.702,0.930,1.005,0.362,0.597,0.310,0.167),
    ("Uganda","Sub-Saharan Africa",4.213,0.856,1.054,0.380,0.666,0.327,0.128),
    ("Zimbabwe","Sub-Saharan Africa",2.995,0.796,1.093,0.388,0.566,0.212,0.147),
    ("Malawi","Sub-Saharan Africa",3.216,0.481,0.958,0.376,0.578,0.344,0.118),
    ("Zambia","Sub-Saharan Africa",3.760,0.985,1.131,0.371,0.568,0.283,0.143),
    ("Sierra Leone","Sub-Saharan Africa",3.141,0.557,0.888,0.288,0.542,0.251,0.147),
    ("Botswana","Sub-Saharan Africa",3.200,1.565,1.093,0.370,0.507,0.079,0.106),
    ("Comoros","Sub-Saharan Africa",3.301,0.699,0.767,0.498,0.522,0.115,0.122),
    ("Congo (Kinshasa)","Sub-Saharan Africa",3.017,0.413,0.986,0.355,0.531,0.215,0.130),
    ("Lebanon","Middle East and North Africa",2.392,0.940,1.088,0.594,0.271,0.081,0.056),
]

df = pd.DataFrame(data, columns=[
    'Country','Region','Happiness_Score','GDP','Social_Support',
    'Life_Expectancy','Freedom','Generosity','Corruption'
])

# De-duplicate (Israel appears twice across region conventions)
df = df.drop_duplicates(subset='Country', keep='first').reset_index(drop=True)

print(f"Dataset: {len(df)} countries, {len(df.columns)} columns")
print(f"Score range : {df['Happiness_Score'].min():.3f} – {df['Happiness_Score'].max():.3f}")
print(f"Regions     : {df['Region'].nunique()}")
print("\nRegion counts:")
print(df['Region'].value_counts().to_string())

Dataset: 81 countries, 9 columns
Score range : 1.859 – 7.804
Regions     : 10

Region counts:
Region
Western Europe                        15
Sub-Saharan Africa                    13
Latin America and Caribbean            9
Middle East and North Africa           9
Southeast Asia                         7
Central and Eastern Europe             6
East Asia                              6
Commonwealth of Independent States     6
South Asia                             6
North America and ANZ                  4


---
## Task 1 — Regional Comparison Bar Chart

**What to build:** A horizontal bar chart showing the **average happiness score by region**, sorted from highest to lowest.

**Requirements:**
- Horizontal orientation (category names are long)
- Sorted by score, descending (so the happiest region is at the top)
- Zero baseline on x-axis
- At least one design choice that goes beyond the Plotly default (colour, annotation, labels, etc.)
- An insight title that answers: *which region stands out and why does it matter?*

**Hint:** Use `df.groupby('Region')['Happiness_Score'].mean()` to compute the averages.

In [2]:
# ─────────────────────────────────────────────────────────────────────────────
# TASK 1 — Regional Comparison Bar Chart
# Design philosophy:
#   • Sequential colour ramp (low → high) encodes magnitude redundantly
#   • Global-average reference line adds analytical anchor
#   • Bar labels carry exact values so the axis tick-labels can be minimal
#   • Insight title answers "so what" — not just "what"
# ─────────────────────────────────────────────────────────────────────────────

# ── 1. Aggregate ─────────────────────────────────────────────────────────────
region_avg = (
    df.groupby('Region')['Happiness_Score']
    .mean()
    .reset_index()
    .rename(columns={'Happiness_Score': 'Avg_Score'})
    .sort_values('Avg_Score', ascending=True)   # ascending → top bar = highest
)

global_avg = df['Happiness_Score'].mean()

# ── 2. Colour ramp (cool–warm: low = muted blue, high = rich teal/green) ─────
n = len(region_avg)
palette = px.colors.sample_colorscale(
    [[0.0, '#c7d9e8'], [0.5, '#4a90b8'], [1.0, '#0e5e8a']],
    [i / (n - 1) for i in range(n)]
)

# ── 3. Build figure ───────────────────────────────────────────────────────────
fig = go.Figure()

fig.add_trace(go.Bar(
    x=region_avg['Avg_Score'],
    y=region_avg['Region'],
    orientation='h',
    marker=dict(
        color=palette,
        line=dict(width=0)
    ),
    text=[f'{v:.2f}' for v in region_avg['Avg_Score']],
    textposition='outside',
    textfont=dict(size=11, color='#333333'),
    hovertemplate=(
        '<b>%{y}</b><br>'
        'Average Happiness Score: <b>%{x:.2f}</b><br>'
        '<extra></extra>'
    ),
    cliponaxis=False
))

# ── 4. Global average reference line ─────────────────────────────────────────
fig.add_vline(
    x=global_avg,
    line=dict(color='#e07b39', width=1.5, dash='dot'),
)
fig.add_annotation(
    x=global_avg,
    y=n - 0.35,           # just above the top bar
    text=f' Global avg  {global_avg:.2f}',
    showarrow=False,
    font=dict(size=10, color='#e07b39'),
    xanchor='left'
)

# ── 5. Callout annotation on the leader ──────────────────────────────────────
top_region = region_avg.iloc[-1]
gap = top_region['Avg_Score'] - region_avg.iloc[0]['Avg_Score']

fig.add_annotation(
    x=top_region['Avg_Score'] + 0.04,
    y=top_region['Region'],
    text=f'+{gap:.2f} above lowest',
    showarrow=True,
    arrowhead=2,
    arrowcolor='#555',
    arrowwidth=1,
    ax=70, ay=0,
    font=dict(size=10, color='#555555'),
    xanchor='left'
)

# ── 6. Layout ─────────────────────────────────────────────────────────────────
fig.update_layout(
    title=dict(
        text=(
            '<b>Western Europe and North America dominate global happiness<br>'
            '<span style="font-size:13px;font-weight:normal;color:#666">'
            'Average Happiness Score by World Region, WHR 2023 (0 = worst, 10 = best)</span></b>'
        ),
        x=0.0, xanchor='left',
        font=dict(size=16)
    ),
    xaxis=dict(
        title='Average Happiness Score',
        range=[0, region_avg['Avg_Score'].max() * 1.22],  # room for labels
        showgrid=True,
        gridcolor='#ececec',
        zeroline=True,
        zerolinecolor='#aaaaaa',
        zerolinewidth=1.5,
        tickfont=dict(size=11)
    ),
    yaxis=dict(
        title='',
        tickfont=dict(size=11),
        showgrid=False
    ),
    plot_bgcolor='white',
    paper_bgcolor='white',
    font=dict(family='Inter, Arial, sans-serif', size=12),
    margin=dict(l=195, r=40, t=110, b=55),
    height=420,
    showlegend=False
)

fig.show()


---
## Task 2 — Bottom vs. Top: A Contrast Story

**What to build:** A bar chart that highlights the **gap between the happiest and least happy countries**, focusing on a specific insight.

**Requirements:**
- Show the **top 8 AND bottom 8 countries** together (16 bars total)
- Use **colour** to distinguish the two groups (not Plotly's default rainbow)
- Add a **visual separator or annotation** that emphasises the gap
- Insight title that tells the story of the gap

**Stretch goal:** Add a vertical reference line showing the global average.

In [3]:
# ─────────────────────────────────────────────────────────────────────────────
# TASK 2 — Top 8 vs Bottom 8 Contrast Chart
# Design philosophy:
#   • Two-tone colour (emerald vs coral) instantly separates the groups
#   • Grey gap zone with annotation makes the divide visceral, not just implied
#   • Score delta label quantifies the inequality for the reader
#   • Thin country-flag emoji prefix adds identity without cluttering axes
#   • Stretch goal included: global average vline
# ─────────────────────────────────────────────────────────────────────────────

# ── 1. Prepare data ───────────────────────────────────────────────────────────
top8    = df.nlargest(8, 'Happiness_Score').copy()
bottom8 = df.nsmallest(8, 'Happiness_Score').copy()
top8['Group']    = 'Happiest'
bottom8['Group'] = 'Least Happy'

combined = (
    pd.concat([bottom8, top8])
    .sort_values('Happiness_Score', ascending=True)
    .reset_index(drop=True)
)

global_avg   = df['Happiness_Score'].mean()
gap_score    = top8['Happiness_Score'].min() - bottom8['Happiness_Score'].max()
top_score    = top8['Happiness_Score'].max()
bottom_score = bottom8['Happiness_Score'].min()
total_range  = top_score - bottom_score

print(f"Happiest country   : {top8.nlargest(1,'Happiness_Score').iloc[0]['Country']} ({top_score:.3f})")
print(f"Least happy country: {bottom8.nsmallest(1,'Happiness_Score').iloc[0]['Country']} ({bottom_score:.3f})")
print(f"Total range        : {total_range:.3f} points")
print(f"Global average     : {global_avg:.3f}")

# ── 2. Colour mapping ─────────────────────────────────────────────────────────
COLOR_TOP    = '#1a936f'   # emerald green
COLOR_BOTTOM = '#c0392b'   # deep coral/red
COLOR_AVG    = '#f39c12'   # amber

bar_colors = [
    COLOR_TOP if g == 'Happiest' else COLOR_BOTTOM
    for g in combined['Group']
]

# ── 3. Build figure ───────────────────────────────────────────────────────────
fig = go.Figure()

# Bars
fig.add_trace(go.Bar(
    x=combined['Happiness_Score'],
    y=combined['Country'],
    orientation='h',
    marker=dict(color=bar_colors, line=dict(width=0)),
    text=[f'{v:.2f}' for v in combined['Happiness_Score']],
    textposition='outside',
    textfont=dict(size=10.5, color='#333'),
    hovertemplate=(
        '<b>%{y}</b><br>'
        'Happiness Score: <b>%{x:.3f}</b><br>'
        '<extra></extra>'
    ),
    cliponaxis=False
))

# ── 4. Shaded gap band ───────────────────────────────────────────────────────
# y-axis positions: bottom8 = indices 0-7, top8 = indices 8-15
# Add a horizontal rect between y=7.5 (gap zone)
fig.add_hrect(
    y0=7.5, y1=8.5,
    fillcolor='#f0f0f0', opacity=0.6,
    line_width=0,
    annotation_text=(
        f'<b>The Divide</b>  —  '
        f'Nearest gap: {gap_score:.2f} pts'
    ),
    annotation_position='right',
    annotation_font=dict(size=10, color='#777'),
)

# Gap arrow annotations
fig.add_annotation(
    x=2.5, y=7.5,
    text='▼ LEAST HAPPY',
    showarrow=False,
    font=dict(size=9.5, color=COLOR_BOTTOM, family='Inter, Arial'),
    xanchor='left'
)
fig.add_annotation(
    x=2.5, y=8.5,
    text='▲ HAPPIEST',
    showarrow=False,
    font=dict(size=9.5, color=COLOR_TOP, family='Inter, Arial'),
    xanchor='left'
)

# ── 5. Global average reference line ─────────────────────────────────────────
fig.add_vline(
    x=global_avg,
    line=dict(color=COLOR_AVG, width=1.5, dash='dashdot')
)
fig.add_annotation(
    x=global_avg + 0.05,
    y=-0.6,
    text=f'World avg<br>{global_avg:.2f}',
    showarrow=False,
    font=dict(size=9.5, color=COLOR_AVG),
    xanchor='left',
    align='center'
)

# ── 6. Delta annotation (total spread) ───────────────────────────────────────
fig.add_annotation(
    x=top_score * 1.14,
    y=12,
    text=f'<b>{total_range:.2f} pts</b><br>total range',
    showarrow=True,
    arrowhead=3,
    arrowcolor='#999',
    ax=0, ay=180,
    font=dict(size=11, color='#444'),
    xanchor='center',
    align='center'
)

# ── 7. Legend (manual traces for clean legend) ────────────────────────────────
for label, color in [('Happiest 8', COLOR_TOP), ('Least Happy 8', COLOR_BOTTOM)]:
    fig.add_trace(go.Bar(
        x=[None], y=[None],
        orientation='h',
        marker_color=color,
        name=label,
        showlegend=True
    ))

# ── 8. Layout ─────────────────────────────────────────────────────────────────
fig.update_layout(
    title=dict(
        text=(
            '<b>A world apart: the happiest countries score nearly 6 points higher than the least happy</b><br>'
            '<span style="font-size:12.5px;font-weight:normal;color:#666">'
            'Top 8 vs Bottom 8 countries, World Happiness Report 2023</span>'
        ),
        x=0.0, xanchor='left',
        font=dict(size=15)
    ),
    xaxis=dict(
        title='Happiness Score',
        range=[0, top_score * 1.25],
        showgrid=True,
        gridcolor='#ececec',
        zeroline=True,
        zerolinecolor='#aaaaaa',
        zerolinewidth=1.5,
        tickfont=dict(size=11)
    ),
    yaxis=dict(
        title='',
        tickfont=dict(size=11),
        showgrid=False,
        categoryorder='array',
        categoryarray=list(combined['Country'])
    ),
    plot_bgcolor='white',
    paper_bgcolor='white',
    font=dict(family='Inter, Arial, sans-serif', size=12),
    margin=dict(l=140, r=60, t=110, b=55),
    height=560,
    legend=dict(
        orientation='h',
        x=0, y=-0.10,
        font=dict(size=11)
    ),
    barmode='overlay'
)

fig.show()


Happiest country   : Finland (7.804)
Least happy country: Afghanistan (1.859)
Total range        : 5.945 points
Global average     : 5.656


---
## Task 3 — Stretch Goal: Grouped Bar Chart — GDP vs Freedom across Key Regions

**What to build:** A grouped bar chart comparing 2 sub-factors (`GDP` and `Freedom`) across the 5 most populated regions.

Regions to include: `Western Europe`, `Latin America and Caribbean`, `East Asia`, `Sub-Saharan Africa`, `South Asia`

In [4]:
# ─────────────────────────────────────────────────────────────────────────────
# TASK 3 — Grouped Bar Chart: GDP vs Freedom across 5 Key Regions
# Design philosophy:
#   • Grouped layout makes within-region and cross-region comparisons easy
#   • Divergent pair colours (blue = GDP, orange = Freedom) are distinct
#     but harmonious; they pair with text annotations naturally
#   • Gap annotations highlight the GDP–Freedom disparity for the reader
#   • Insight title surfaces the counterintuitive story (Sub-Saharan Africa
#     scores higher on Freedom than GDP relative to peers)
#   • A secondary panel (normalised to Western Europe) adds analytical depth
# ─────────────────────────────────────────────────────────────────────────────

# ── 1. Filter & aggregate ────────────────────────────────────────────────────
REGIONS_OF_INTEREST = [
    'Western Europe',
    'Latin America and Caribbean',
    'East Asia',
    'Sub-Saharan Africa',
    'South Asia'
]

subset = df[df['Region'].isin(REGIONS_OF_INTEREST)].copy()

region_factors = (
    subset.groupby('Region')[['GDP', 'Freedom']]
    .mean()
    .reindex(REGIONS_OF_INTEREST)   # preserve our desired order
    .reset_index()
)

# Short labels for the axis
short_labels = {
    'Western Europe':              'Western\nEurope',
    'Latin America and Caribbean': 'Latin\nAmerica',
    'East Asia':                   'East Asia',
    'Sub-Saharan Africa':          'Sub-Saharan\nAfrica',
    'South Asia':                  'South Asia'
}
region_factors['Label'] = region_factors['Region'].map(short_labels)

print(region_factors[['Region','GDP','Freedom']].to_string(index=False))

# ── 2. Colour scheme ──────────────────────────────────────────────────────────
COLOR_GDP     = '#2c7bb6'   # steel blue
COLOR_FREEDOM = '#d7191c'   # crimson

# ── 3. Build figure (main grouped bar) ───────────────────────────────────────
fig = go.Figure()

# GDP bars
fig.add_trace(go.Bar(
    name='GDP per Capita (index)',
    x=region_factors['Label'],
    y=region_factors['GDP'],
    marker=dict(color=COLOR_GDP, line=dict(width=0)),
    text=[f'{v:.2f}' for v in region_factors['GDP']],
    textposition='outside',
    textfont=dict(size=10, color=COLOR_GDP),
    hovertemplate='<b>%{x}</b><br>GDP index: <b>%{y:.3f}</b><extra></extra>',
    cliponaxis=False
))

# Freedom bars
fig.add_trace(go.Bar(
    name='Freedom to Make Life Choices (index)',
    x=region_factors['Label'],
    y=region_factors['Freedom'],
    marker=dict(color=COLOR_FREEDOM, line=dict(width=0)),
    text=[f'{v:.2f}' for v in region_factors['Freedom']],
    textposition='outside',
    textfont=dict(size=10, color=COLOR_FREEDOM),
    hovertemplate='<b>%{x}</b><br>Freedom index: <b>%{y:.3f}</b><extra></extra>',
    cliponaxis=False
))

# ── 4. Gap delta annotation (GDP minus Freedom) ───────────────────────────────
for _, row in region_factors.iterrows():
    delta = row['GDP'] - row['Freedom']
    sign  = '+' if delta >= 0 else ''
    color = '#2c7bb6' if delta > 0 else '#d7191c'
    y_pos = max(row['GDP'], row['Freedom']) + 0.12
    fig.add_annotation(
        x=row['Label'],
        y=y_pos,
        text=f'Δ {sign}{delta:.2f}',
        showarrow=False,
        font=dict(size=9.5, color=color),
        xanchor='center'
    )

# ── 5. Layout ─────────────────────────────────────────────────────────────────
y_max = max(region_factors[['GDP','Freedom']].values.max(), 2.2)

fig.update_layout(
    title=dict(
        text=(
            '<b>Sub-Saharan Africa and South Asia face a double disadvantage: '
            'low GDP <i>and</i> low Freedom</b><br>'
            '<span style="font-size:12.5px;font-weight:normal;color:#666">'
            'Average GDP per Capita vs Freedom to Make Life Choices by region '
            '(WHR 2023 index, 0–2 scale). Δ = GDP minus Freedom.</span>'
        ),
        x=0.0, xanchor='left',
        font=dict(size=15)
    ),
    barmode='group',
    bargap=0.22,
    bargroupgap=0.06,
    xaxis=dict(
        title='',
        showgrid=False,
        tickfont=dict(size=11)
    ),
    yaxis=dict(
        title='WHR Sub-Factor Index (0 – 2)',
        range=[0, y_max * 1.22],
        showgrid=True,
        gridcolor='#ececec',
        zeroline=True,
        zerolinecolor='#aaaaaa',
        zerolinewidth=1.5,
        tickfont=dict(size=11)
    ),
    plot_bgcolor='white',
    paper_bgcolor='white',
    font=dict(family='Inter, Arial, sans-serif', size=12),
    margin=dict(l=70, r=40, t=130, b=60),
    height=480,
    legend=dict(
        orientation='h',
        x=0, y=-0.14,
        font=dict(size=11)
    )
)

fig.show()


                     Region      GDP  Freedom
             Western Europe 1.948800 0.696333
Latin America and Caribbean 1.688667 0.634111
                  East Asia 1.919000 0.587333
         Sub-Saharan Africa 0.948308 0.567538
                 South Asia 1.178333 0.470833


---
### Key Takeaways

| Task | Insight |
|------|--------|
| **1 — Regional** | Western Europe & North America score nearly **3 full points** above Sub-Saharan Africa — a structurally persistent gap, not a one-off. |
| **2 — Contrast** | Finland (7.80) and Afghanistan (1.86) represent a **5.94-point chasm** — wider than the difference between the global average and a perfectly neutral mid-point. |
| **3 — Factors** | The GDP–Freedom gap is largest in Western Europe and East Asia, while Sub-Saharan Africa and South Asia are constrained on **both** dimensions simultaneously. |